### PR TITLE
Add functionality to allow addition of new value from yaml

### PR DESCRIPTION
**Pull request by Rizhiy**

### CODE REVIEW

**Comment by rbgirshick** [comment]
This PR seems generally useful, thanks!

But I'm unsure of one aspect of the design. Can you give some more context on why you want to prevent adding new keys to a `CfgNode` when that node has a child `CfgNode`?

For example, in your test case you have:

```python
cfg.KWARGS = CN(new_allowed=True)
cfg.KWARGS.z = 0
```

and to this, one can inject a new key, e.g. `KWARGS.a`, if it's of a valid primitive type.

But if I have

```python
cfg.KWARGS = CN(new_allowed=True)
cfg.KWARGS.z = 0
cfg.KWARGS.q = CN()
cfg.KWARGS.q.foo = "123"
```

then I cannot inject `KWARGS.a` and in fact this config fails on construction.

1. It seems generally more useful to allow for adding new keys even when a child `CfgNode` exists.
2. It also seems like you might want to enable adding new child `CfgNode`s if the parent has `new_allowed` set to true (and these new child nodes should have `new_allowed=True` as well).

**Comment by Rizhiy** [comment]
> then I cannot inject KWARGS.a and in fact this config fails on construction.

Actually, you should be able to inject a new key, but as you say it fails on construction.

> Can you give some more context on why you want to prevent adding new keys to a CfgNode when that node has a child CfgNode?

I think that adding `CfgNode`s from yaml is not necessarily a good idea, since we can't check that new keys have required type and might lead to subtle bugs (e.g. wrong indentation in yaml file) which would be difficult to figure out.

**Comment by Rizhiy** [comment]
I think that you are right, we should let the end user decide how to organise the config.
Therefore, I removed the restriction from new values.
Now new nested `CfgNode`s can be added from yaml.

File: yacs/config.py [hunk 1]
```diff
@@ -383,7 +389,7 @@ def _merge_a_into_b(a, b, root, key_list):
     for k, v_ in a.items():
         full_key = ".".join(key_list + [k])
```
Comment (rbgirshick) [inline yacs/config.py]: I think the logic in this loop has gotten a bit out of hand and L406 is too difficult to reason about. There are four cases to deal with as outlined in this simplified code sketch:

```python
if k in b:
    # Either recursively merge (if v is a CfgNode) or update in place
    if isinstance(v, CfgNode):
        _merge_a_into_b(v, b[k], root, key_list + [k])
    else:
        b[k] = v
else:
    # Either add new key inplace or raise an error (unless deprecated)
    if b.is_new_allowed():
        b[k] = v
    else:
        ...  # continue if deprecated else raise error
```

Before merging, I think it's necessary to refactor to make these cases more clear and remove the complicated logic.

Comment (Rizhiy) [reply yacs/config.py]: I have simplified logic, hope it is better now.

Comment (rbgirshick) [reply yacs/config.py]: Yes, this is great. Thanks!

**Comment by rbgirshick** [comment]
Thanks for the the back-and-forth refinement. I'll merge now.

**Pull request merged**